artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPConnectionContext.java

```diff
@@ -416,42 +414,57 @@ public void onRemoteOpen(Connection connection) throws Exception {
        * */
       if (connection.getRemoteProperties() == null || !connection.getRemoteProperties().containsKey(CONNECTION_OPEN_FAILED)) {
          long nextKeepAliveTime = handler.tick(true);
-         synchronized (schedulingLock) {
-            if (nextKeepAliveTime != 0 && scheduledPool != null && !isSchedulingCancelled) {
-               scheduledFuture = scheduledPool.schedule(new ScheduleRunnable(), (nextKeepAliveTime - TimeUnit.NANOSECONDS.toMillis(System.nanoTime())), TimeUnit.MILLISECONDS);
-            }
+
+         if (nextKeepAliveTime != 0 && scheduledPool != null) {
+            scheduleOp.setDelay(nextKeepAliveTime - TimeUnit.NANOSECONDS.toMillis(System.nanoTime()));
+
+            scheduledFutureRef.getAndUpdate(scheduleOp);
```

Comment:
there is some specific return value from getOrUpdate that you expect to find? if yes, maybe you can put an assert there to help spot issues on tests...

Reply:
The return value depends on current value so I don't know how to predict it without access to scheduledFutureRef.